### DISCUSSION TITLE
Multichain API

### DISCUSSION THREAD

**Discussion by vandan** [MetaMask/metamask-improvement-proposals#53] [MIPs]
We, as MetaMask API Maintainers, intend to introduce a new [Wallet API](https://docs.metamask.io/wallet/concepts/wallet-api/) that is designed for simultaneous interactions with EVM and non-EVM RPC networks available through the wallet. We refer it as the _Multichain API_. The new interface is based on open standards that we have contributed to as part of our participation in [Chain Agnostic Standards Alliance](https://github.com/ChainAgnostic). 

The Multichain API is designed around the following standards:

- API lifecycle overview [CAIP-316](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-316.md)
- wallet_createSession [CAIP-25](https://chainagnostic.org/CAIPs/caip-25)
- wallet_sessionChanged [CAIP-311](https://chainagnostic.org/CAIPs/caip-311)
- wallet_getSession [CAIP-312](https://chainagnostic.org/CAIPs/caip-312)
- wallet_revokeSession [CAIP-285](https://chainagnostic.org/CAIPs/caip-285)
- wallet_invokeMethod [CAIP-27](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-27.md)
- wallet_notify [CAIP-319](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-319.md)

There is also a [draft OpenRPC specification](https://github.com/MetaMask/api-specs/blob/main/multichain/openrpc.yaml) available for the Multichain API.

We hope these resources are helpful to track the latest developments on specifications and standards that will guide the implementation. Of course, additional MIPs may be posted to provide more detail and guidelines about what is coming.

We look forward to hearing any feedback on our plans to create an implementation that is based on these standards. Thank you!

**Comment by vandan** [comment]
We have posted an [initial PR for MIPs](https://github.com/MetaMask/metamask-improvement-proposals/pull/54) with additional background and implementation-specific nuances related to the Multichain API.

**Comment by vandan** [comment]
Maintainers have posted MIPs under `Draft` status for the planned Multichain API:
[MIP-5](https://github.com/MetaMask/metamask-improvement-proposals/blob/main/MIPs/mip-5.md)
[MIP-6](https://github.com/MetaMask/metamask-improvement-proposals/blob/main/MIPs/mip-6.md)

Please review and share any feedback or concerns here.

Unless any significant feedback is received, we expect these MIPs to be moved to `Review` status shortly after a few outstanding details have been added.

**Comment by vandan** [comment]
Adding a reference to a [PR for CAIP-282](https://github.com/ChainAgnostic/CAIPs/pull/282) for the latest on standardized messaging for the wallet interfaces in browser environments, which is intended to be incorporated into MetaMask's Multichain API.

**Comment by MicahZoltu** [comment]
It feels like if MetaMask is going to be implementing an entirely new API for communication between apps and wallets, this feels like the perfect opportunity to *also* change the communication channel to `window.postMessage` instead of an injected script.  This would allow bi-directional communication that doesn't require the extension have superuser permissions, and it also enables things like iframe based wallets (e.g., Gnosis SAFE style) rather than only working with extension based wallets.

The reason this is a good opportunity timing wise is because anyone who wants to utilize this new wallet API must implement new libraries/code (it isn't backward compatible), so it is best to do multiple breaking changes at once rather than implementing all of this and then having another breaking change down the line where the communication channel changes.

I'm aware that this is somewhat separate from the API definition itself, but I am not sure where else to provide this feedback.
